xlators/cluster/dht/src/dht-inode-write.c

```diff
@@ -303,415 +161,32 @@ dht_truncate_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int op_ret,
-                      subvol->fops->fallocate, local->fd,
-                      local->rebalance.flags, local->rebalance.offset,
-                      local->rebalance.size, local->xattr_req);
-
-    return 0;
-
-out:
-    DHT_STACK_UNWIND(fallocate, frame, -1, op_errno, NULL, NULL, NULL);
-    return 0;
-}
-
-int
-dht_fallocate(call_frame_t *frame, xlator_t *this, fd_t *fd, int32_t mode,
-              off_t offset, size_t len, dict_t *xdata)
-{
-    xlator_t *subvol = NULL;
-    int op_errno = -1;
-    dht_local_t *local = NULL;
-
-    VALIDATE_OR_GOTO(frame, err);
-    VALIDATE_OR_GOTO(this, err);
-    VALIDATE_OR_GOTO(fd, err);
-
-    local = dht_local_init(frame, NULL, fd, GF_FOP_FALLOCATE);
-    if (!local) {
-        op_errno = ENOMEM;
-        goto err;
-    }
-
-    local->rebalance.flags = mode;
-    local->rebalance.offset = offset;
-    local->rebalance.size = len;
-
-    local->call_cnt = 1;
-    subvol = local->cached_subvol;
-    if (!subvol) {
-        gf_msg_debug(this->name, 0, "no cached subvolume for fd=%p", fd);
-        op_errno = EINVAL;
-        goto err;
-    }
-
-    if (xdata)
-        local->xattr_req = dict_ref(xdata);
-
-    STACK_WIND_COOKIE(frame, dht_fallocate_cbk, subvol, subvol,
-                      subvol->fops->fallocate, fd, local->rebalance.flags,
-                      local->rebalance.offset, local->rebalance.size,
-                      local->xattr_req);
-
-    return 0;
-
-err:
-    op_errno = (op_errno == -1) ? errno : op_errno;
-    DHT_STACK_UNWIND(fallocate, frame, -1, op_errno, NULL, NULL, NULL);
-
-    return 0;
-}
-
-int
-dht_discard_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int op_ret,
-                int op_errno, struct iatt *prebuf, struct iatt *postbuf,
-                dict_t *xdata)
-{
-    dht_local_t *local = NULL;
-    xlator_t *prev = NULL;
-    int ret = -1;
-    xlator_t *src_subvol = NULL;
-    xlator_t *dst_subvol = NULL;
-
-    GF_VALIDATE_OR_GOTO("dht", frame, err);
-    GF_VALIDATE_OR_GOTO("dht", this, out);
-    GF_VALIDATE_OR_GOTO("dht", frame->local, out);
-    GF_VALIDATE_OR_GOTO("dht", cookie, out);
-
-    local = frame->local;
-    prev = cookie;
-
-    /* discard fails with EBADF if dht has not yet opened the fd
-     * on the cached subvol. This could happen if the file was migrated
-     * and a lookup updated the cached subvol in the inode ctx.
-     * We only check once as this could actually be a valid error.
-     */
-    if (dht_check_remote_fd_failed_error(local, op_ret, op_errno)) {
-        ret = dht_check_and_open_fd_on_subvol(this, frame);
-        if (ret)
+    /* Check if the rebalance phase1 is true.
+       At the end of the migration process, whatever 'attr' we
+       have on source file will be migrated to destination file
+       in one shot, hence we don't need to check for in progress
+       state here (ie, PHASE1) for [f]setattr*/
+    if ((local->fop != GF_FOP_SETATTR) && (local->fop != GF_FOP_SETATTR) &&
```

Comment:
I don't understand this if statement, two first conditions look identical to me.

Reply:
You are correct, the next one should have been GF_FOP_FSETATTR. Let me see if I can automate test for this fop too using gfapi.